openlibrary/accounts/model.py

```diff
@@ -597,42 +604,90 @@ def audit_accounts(email, password, test=False):
+        # account...
         if not link:
+            # then check if there's an Open Library account which shares
+            # the same email as this IA account.
             ol_account = OpenLibraryAccount.get(email=email, test=test)
-            link = ia_account.itemname if ol_account else None
 
-        # So long as there's either a linked OL account, or an OL
+            # If an Open Library account with a matching email account exist...
+            if ol_account:
+                # Check whether it is linked already, i.e. has an itemname
+                # set. We already determined that no OL account is
+                # linked to our IA account. Therefore this Open
+                # Library account having the same email as our IA
+                # account must have been linked to a different
+                # Internet Archive account.
+                if ol_account.itemname:
+                    return {'error': 'wrong_ia_account'}
+
+        # At this point, it must either be the case that (a)
+        # `ol_account` already links to our IA account (in which case
+        # `link` has a correct value), (b) that an unlinked
+        # `ol_account` shares the same email as our IA account and
+        # thus can and should be safely linked to our IA account, or
+        # (c) no `ol_account` which is linked or can be linked has
+        # been found and therefore, asuming
+        # lending.config_ia_auth_only is enabled, we need to create
+        # and link it.
+        if not ol_account:
+            if lending.config_ia_auth_only:
+                try:
+                    ol_account = OpenLibraryAccount.create(
+                        ia_account.itemname, email, password,
+                        displayname=ia_account.screenname,
+                        verified=True, test=test)
+                except ValueError as e:
+                    return {'error': 'max_retries_exceeded', 'msg': str(e)}
+
+                if not ol_account.itemname:
+                    ol_account.link(ia_account.itemname)
+                    stats.increment('ol.account.xauth.ia-auto-created-ol')
+
+                audit['link'] = ia_account.itemname
+                audit['has_ol'] = ol_account.email
+                web.ctx.conn.set_auth_token(ol_account.generate_login_code())
+
+        # So long as there's either a linked OL account, or an unlinked OL
         # account with the same email, set them as linked (and let the
-        # finalize logic link them, if necessary)
-        if ol_account:
+        # finalize logic link them, if needed)
+        else:
+            if lending.config_ia_auth_only:
+                if not ol_account.itemname:
+                    ol_account.link(ia_account.itemname)
+                    stats.increment('ol.account.xauth.auto-linked')
             if not ol_account.verified:
-                return {'error': 'account_not_verified'}
+                if lending.config_ia_auth_only:
```

Comment:
I think this is about the same level of complexity in either case as `ol_account.verified` would then have to be checked in 2 locations instead of lending.config_ia_auth_only